src/main/java/slimeknights/tconstruct/smeltery/tileentity/FaucetTileEntity.java

```diff
@@ -0,0 +1,231 @@
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.fluids.capability.CapabilityFluidHandler;
+import net.minecraftforge.fluids.capability.IFluidHandler;
+import slimeknights.tconstruct.smeltery.TinkerSmeltery;
+
+import javax.annotation.Nullable;
+
+import static slimeknights.tconstruct.smeltery.block.FaucetBlock.FACING;
+
+public class FaucetTileEntity extends TileEntity implements ITickableTileEntity {
+  public Direction direction;
+  public boolean isPouring;
+  public boolean stopPouring;
+  public FluidStack drained;
+  public boolean lastRedstoneState;
+
+  public FaucetTileEntity() {
+    this(TinkerSmeltery.faucet.get());
+  }
+  public FaucetTileEntity(TileEntityType<?> tileEntityTypeIn) {
+    super(tileEntityTypeIn);
+    reset();
+  }
+
+  public void activate() {
+    BlockState state = getWorld().getBlockState(pos);
+    if (!state.getProperties().contains(FACING)) {
+      return;
+    }
+
+    // already pouring? we want to stop then
+    if (isPouring) {
+      stopPouring = true;
+      return;
+    }
+
+    direction = getWorld().getBlockState(pos).get(FACING);
+    doTransfer();
+    return;
+  }
+
+  public void handleRedstone(boolean hasSignal) {
+    if (hasSignal != lastRedstoneState) {
+      lastRedstoneState = hasSignal;
+      if (hasSignal) {
+        getWorld().getPendingBlockTicks().scheduleTick(pos, this.getBlockState().getBlock(), 2);
+      }
+    }
+  }
+
+  @Override
+  public void tick() {
+    if (getWorld().isRemote) {
+      return;
+    }
+
+    // nothing to do if not pouring
+    if (!isPouring) {
+      return;
+    }
+
+    if (drained != null) {
+      // done draining
+      if (drained.getAmount() <= 0) {
+        drained = null;
+        // pour me another, if we want to
+        if (!stopPouring) {
+          doTransfer();
+        }
+        else {
+          reset();
+        }
+      }
+      else {
+        // reduce amount (cooldown)
+        pour();
+      }
+    }
+  }
+
+  protected void doTransfer() {
+    // still got content left
+    IFluidHandler toDrain = getFluidHandler(pos.offset(direction), direction.getOpposite());
+    IFluidHandler toFill = getFluidHandler(pos.down(), Direction.UP);
+    if (toDrain != null && toFill != null) {
+      // can we drain?
+      FluidStack drained = toDrain.drain(144, IFluidHandler.FluidAction.SIMULATE);
+      if (!drained.isEmpty()) {
+        // can we fill
+        int filled = toFill.fill(drained, IFluidHandler.FluidAction.SIMULATE);
+        if (filled > 0) {
+          // drain the liquid and transfer it, buffer the amount for delay
+          this.drained = toDrain.drain(filled, IFluidHandler.FluidAction.EXECUTE);
+          this.isPouring = true;
+          pour();
+
+          // sync to clients
+          if (!getWorld().isRemote()) { // TODO: getWorld() instanceof WorldServer
+            // TODO: TinkerNetwork.sendToClients((WorldServer) getWorld(), pos, new FaucetActivationPacket(pos, drained));
```

Comment:
Why is this commented out? You probably want `ServerWorld`

Reply:
Ah, because I wasn't sure at the time. Do we currently have a faucet activation
packet, or do I need to write it?

Reply:
We had one in 1.12. You need to port it.

Reply:
Think I did this right, however, faucet'ing fluids between two tanks
will spam the logs with:
```[Render thread/WARN] [minecraft/ClientPlayNetHandler]: Unknown custom packet identifier: minecraft:tconstruct```
I am probably missing something simple.

Reply:
looks like I have to setPacketHandled.